pkg/syncer/mysqlcluster/statefullset.go

```diff
@@ -0,0 +1,675 @@
+
+	dataVolumeName = "data"
+	// DataVolumeMountPath is the path to mysql data
+	DataVolumeMountPath = "/var/lib/mysql"
+)
+
+type sfsSyncer struct {
+	cluster           *clusterwrap.MysqlCluster
+	configMapRevision string
+	secretRevision    string
+	opt               *options.Options
+}
+
+// NewStatefulSetSyncer returns a syncer for stateful set
+func NewStatefulSetSyncer(cluster *api.MysqlCluster, cmRev, secRev string, opt *options.Options) syncer.Interface {
+	return &sfsSyncer{
+		cluster:           clusterwrap.NewMysqlClusterWrapper(cluster),
+		configMapRevision: cmRev,
+		secretRevision:    secRev,
+		opt:               opt,
+	}
+}
+
+func (s *sfsSyncer) GetExistingObjectPlaceholder() runtime.Object {
+	return &apps.StatefulSet{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      s.cluster.GetNameForResource(api.StatefulSet),
+			Namespace: s.cluster.Namespace,
+		},
+	}
+}
+
+func (s *sfsSyncer) ShouldHaveOwnerReference() bool {
+	return true
+}
+
+func (s *sfsSyncer) Sync(in runtime.Object) error {
+	out := in.(*apps.StatefulSet)
+
+	if out.Status.ReadyReplicas == out.Status.Replicas {
+		s.cluster.UpdateStatusCondition(api.ClusterConditionReady,
+			core.ConditionTrue, "statefulset ready", "Cluster is ready.")
+	} else {
+		s.cluster.UpdateStatusCondition(api.ClusterConditionReady,
+			core.ConditionFalse, "statefulset not ready", "Cluster is not ready.")
+	}
+
+	s.cluster.Status.ReadyNodes = int(out.Status.ReadyReplicas)
+
+	out.Spec.Replicas = &s.cluster.Spec.Replicas
+	out.Spec.Selector = &metav1.LabelSelector{
+		MatchLabels: s.getLabels(map[string]string{}),
+	}
+
+	out.Spec.ServiceName = s.cluster.GetNameForResource(api.HeadlessSVC)
+	out.Spec.Template = s.ensureTemplate(out.Spec.Template)
+	out.Spec.VolumeClaimTemplates = s.ensureVolumeClaimTemplates(out.Spec.VolumeClaimTemplates)
+
+	return nil
+}
+
+func (s *sfsSyncer) ensureTemplate(in core.PodTemplateSpec) core.PodTemplateSpec {
+	in.ObjectMeta.Labels = s.getLabels(s.cluster.Spec.PodSpec.Labels)
+	in.ObjectMeta.Annotations = s.cluster.Spec.PodSpec.Annotations
+	if len(in.ObjectMeta.Annotations) == 0 {
+		in.ObjectMeta.Annotations = make(map[string]string)
+	}
+	in.ObjectMeta.Annotations["config_rev"] = s.configMapRevision
+	in.ObjectMeta.Annotations["secret_rev"] = s.secretRevision
+	in.ObjectMeta.Annotations["prometheus.io/scrape"] = "true"
+	in.ObjectMeta.Annotations["prometheus.io/port"] = fmt.Sprintf("%d", ExporterPort)
+
+	in.Spec.InitContainers = s.ensureInitContainersSpec(in.Spec.InitContainers)
+	in.Spec.Containers = s.ensureContainersSpec(in.Spec.Containers)
+
+	in.Spec.Volumes = s.ensureVolumes(in.Spec.Volumes)
+
+	in.Spec.Affinity = &s.cluster.Spec.PodSpec.Affinity
+	in.Spec.NodeSelector = s.cluster.Spec.PodSpec.NodeSelector
+	in.Spec.ImagePullSecrets = s.cluster.Spec.PodSpec.ImagePullSecrets
+
+	return in
+}
+
+const (
+	containerInitName      = "init-mysql"
+	containerCloneName     = "clone-mysql"
+	containerHelperName    = "helper"
```

Comment:
Let's rename `containerHelperName` -> `containerSidecarName` and `helper` -> `mysql-operator-sidecar`